File: src/lib/scala/Slate.Common/src/main/scala/slate/common/Ioc.scala
/**
<slate_header>
  author: Kishore Reddy
  url: https://github.com/kishorereddy/scala-slate
  copyright: 2016 Kishore Reddy
  license: https://github.com/kishorereddy/scala-slate/blob/master/LICENSE.md
  desc: a scala micro-framework
  usage: Please refer to license on github for more info.
</slate_header>
  */

package slate.common

import scala.collection.mutable.Map

object Ioc {
  private val _lookup = Map[String, Any]()


  def register(key:String, obj:Any): Unit =
  {
    _lookup(key) = obj
  }


  def get(key:String):Option[Any] =
  {
    if(_lookup.contains(key))
      return Some(_lookup(key))
    None
  }
}

File: src/lib/kotlin/slatekit-common/src/main/kotlin/slatekit/common/Ioc.kt
package slatekit.common

object Ioc {
    val lookup = mutableMapOf<String, (String) -> Any>()

    fun register(name: String, creator: (String) -> Any) {
        lookup.put(name, creator)
    }

    fun <T> get(name: String): T {
        if (!lookup.containsKey(name)) {
            throw IllegalArgumentException("Component $name not in Ioc")
        }
        val creator = lookup.get(name)!!
        val instance = creator.invoke(name) as T
        return instance
    }

    fun contains(name: String): Boolean {
       return lookup.containsKey(name)
    }
}